packages/block-library/src/table/edit.js

```diff
@@ -0,0 +1,295 @@
+		this.onChangeInitialRowCount = this.onChangeInitialRowCount.bind( this );
+		this.renderSection = this.renderSection.bind( this );
+		this.getTableControls = this.getTableControls.bind( this );
+		this.createOnDeleteRow = this.createOnDeleteRow.bind( this );
+		this.createOnDeleteColumn = this.createOnDeleteColumn.bind( this );
+		this.createOnInsertRow = this.createOnInsertRow.bind( this );
+		this.createOnInsertColumn = this.createOnInsertColumn.bind( this );
+		this.createOnFocus = this.createOnFocus.bind( this );
+
+		this.state = {
+			initialRowCount: 2,
+			initialColumnCount: 2,
+			selectedCell: null,
+		};
+	}
+
+	onChangeInitialColumnCount( initialColumnCount ) {
+		this.setState( { initialColumnCount } );
+	}
+
+	onChangeInitialRowCount( initialRowCount ) {
+		this.setState( { initialRowCount } );
+	}
+
+	onCreateTable() {
+		const { setAttributes } = this.props;
+		const { initialRowCount, initialColumnCount } = this.state;
+
+		setAttributes( createTable( {
+			rowCount: initialRowCount,
+			columnCount: initialColumnCount,
+		} ) );
+	}
+
+	onChangeFixedLayout() {
+		const { attributes, setAttributes } = this.props;
+		const { hasFixedLayout } = attributes;
+
+		setAttributes( { hasFixedLayout: ! hasFixedLayout } );
+	}
+
+	createOnChange( { section, rowIndex, columnIndex } ) {
+		const { attributes, setAttributes } = this.props;
+
+		return ( content ) => {
+			setAttributes( updateCellContent( attributes, {
+				section,
+				rowIndex,
+				columnIndex,
+				content,
+			} ) );
+		};
+	}
+
+	createOnInsertRow( { section, rowIndex }, delta ) {
+		const { attributes, setAttributes } = this.props;
+
+		rowIndex += delta;
+
+		return () => {
+			setAttributes( insertRow( attributes, { section, rowIndex } ) );
+			this.setState( { selectedCell: null } );
+		};
+	}
+
+	createOnDeleteRow( { section, rowIndex } ) {
+		const { attributes, setAttributes } = this.props;
+
+		return () => {
+			setAttributes( deleteRow( attributes, { section, rowIndex } ) );
+			this.setState( { selectedCell: null } );
+		};
+	}
+
+	createOnInsertColumn( { section, columnIndex }, delta ) {
+		const { attributes, setAttributes } = this.props;
+
+		columnIndex += delta;
+
+		return () => {
+			setAttributes( insertColumn( attributes, { section, columnIndex } ) );
+			this.setState( { selectedCell: null } );
+		};
+	}
+
+	createOnDeleteColumn( { section, columnIndex } ) {
+		const { attributes, setAttributes } = this.props;
+
+		return () => {
+			setAttributes( deleteColumn( attributes, { section, columnIndex } ) );
+			this.setState( { selectedCell: null } );
+		};
+	}
+
+	createOnFocus( selectedCell ) {
+		return () => {
+			this.setState( { selectedCell } );
+		};
+	}
+
+	getTableControls() {
+		const { selectedCell } = this.state;
+
+		return [
+			{
+				icon: 'table-row-before',
+				title: __( 'Add Row Before' ),
+				isDisabled: ! selectedCell,
+				onClick: selectedCell && this.createOnInsertRow( selectedCell ),
```

Comment:
Interesting!

Reply:
You'd still have to create the function though to pass the cell location.

Reply:
Something like this could be a trade-off which solves both the binding and misdirection issues. It means binding more methods in the constructor, but I think it's easier to read:
```
insertRow( delta ) {
    if ( ! this.state.selectedCell ) {
        return;
    }

    const { section, rowIndex } = this.state.selectedCell;
    const { attributes, setAttributes } = this.props;
    const newRowIndex = rowIndex + delta;

    setAttributes( insertRow( attributes, { section, rowIndex: newRowIndex } ) );
    this.setState( { selectedCell: null } );
}

onInsertRowBefore() {
     this.insertRow(0);
}

onInsertRowAfter() {
     this.insertRow(1);
}

getTableControls() {
    return [
        {
	    icon: 'table-row-before',
	    title: __( 'Add Row Before' ),
	    isDisabled: ! selectedCell,
	    onClick: this.onInsertRowBefore,
        },
        //...
    ];
}
```

Anyway, that's just a stylistic thing. I think the bigger issue is the use of a closure to access `selectedCell`, which can be avoided given this.state.selectedCell can be accessed directly in those functions. That could be the cause of the bug mentioned by @noisysocks.

Reply:
plus 💯 to what @talldan shared. In particular, having methods like `onInsertRowBefore` and `onInsertRowAfter` improves readability of the code by replacing indexes with the description of their functionality.